### ISSUE TITLE
Suggestion: Cmd.flatten : Cmd<Cmd<'a>> -> Cmd<'a>

### ISSUE THREAD

**Issue by jwosty** [fable-compiler/fable-elmish#208]
### Description

May I suggest that Elmish have a Cmd.flatten function, which is the classic monadic flatten operation. Here is a complete implementation:
```fs
    let flatten (cmd: Cmd<Cmd<'a>>) : Cmd<'a> =
        [
            for (sub: ((Cmd<'a> -> unit) -> unit)) in cmd do
                (fun (dispatch: ('a -> unit)) ->
                    let dispatchCmd : Dispatch<Cmd<'a>> = fun (msg: Cmd<'a>) ->
                        for sub in msg do
                            sub dispatch
                    sub dispatchCmd
                )
        ]
```
Most recently, it let me mix and match various layers of Cmds and asyncs:
```fs
    let someAsyncFunction x = async { return x + 1 }
    type Msg = | SomeMsg of int | AnotherMsg
    let makeCmd x : Cmd<Msg> =
        let work = async {
            let! result = someAsyncFunction x
            return Cmd.ofSub (fun dispatch ->
                dispatch (SomeMsg result)
                dispatch AnotherMsg
            )
        }
        work |> Cmd.OfAsync.result |> Cmd.flatten
```
I can easily submit a PR for this with the above code; I just wanted to know if you would be likely to take it.

**Comment by forki** [comment]
Is this something that can be found in elm? In many cases this lib is
following elm and if it's already there then you have a good argument to
add it here.

John Wostenberg <notifications@github.com> schrieb am Mi., 29. Juli 2020,
07:37:

**Comment by MangelMaxime** [comment]
@forki I don't think Elm has this kind of helpers in it.

If I understand correctly your example. You are generating new `Commands` from inside a `Command`?

I think the standard approach is in general, to make this happen in the `update` function when receiving the initial success/error `Msg`.

I am mentioning it because I don't know if creating `Commands` inside of a `Command` is a problem or not.

**Comment by jwosty** [comment]
@MangelMaxime you are correct -- however for the case that I based the example off of, it's really more of a situation where I wanted to dispatch multiple messages from inside the async. I could have alternatively written it manually like so (haven't checked this code in a compiler but it should be mostly right):

```fs
    let someAsyncFunction x = async { return x + 1 }
    type Msg = | SomeMsg of int | AnotherMsg
    let makeCmd x : Cmd<Msg> =
        Cmd.ofSub (fun dispatch ->
            Async.Start <| async {
                let! result = someAsyncFunction x
                dispatch (SomeMsg result)
                AnotherMsg
            }
        )
```
But I didn't really like the explicit ``Async.Start``. I wanted to see if I could do it with Cmd functions directly, and it turns out to be impossible. An alternative might be to have some kind of ``Cmd.Async.returnMany`` which would dispatch a list of messages instead of a single message. Actually, having now written this, I'm realizing that I also could have written an "adapter" dispatch function that would just take a ``Msg list``... So I guess there's many different ways to do this same thing. What would be the most Elmish-ish?

**Comment by et1975** [comment]
@jwosty Can you see if this can be reimplemented with `Cmd.batch`?
